### PR TITLE
Optimize FileInfo stats query

### PR DESCRIPTION
**Pull request by agnivade**
We Denormalize Post.ChannelId on FileInfo.ChannelId

```release-note
The file info stats query is now optimized by denormalizing the channelID column into the table itself. This will speed up the query to get the file count for a channel on clicking the RHS.

Migration times:
On a MySQL 8.0.31 DB with
1405 rows in FileInfo and 11M posts, it took around 0.3s

On a Postgres 12.14 DB with
1731 rows in FileInfo and 11M posts, it took around 0.27s
```

### CODE REVIEW

**Comment by agnivade** [comment]
It's finally here :)

@lieut-data - One additional thing that I have done from your PR is to update the upsert query as well. Please see https://github.com/mattermost/mattermost-server/pull/22513/commits/950bc1271d99a06823061d32a3b3e55cd0d92d92. Just ensuring that this was an oversight, and not something you intentionally avoided.

**Comment by agarciamontoro** [comment]
@agnivade, were you able to run load-tests to verify the improvement?

**Comment by agnivade** [comment]
Yep! Will post results later.

**Comment by lieut-data** [comment]
This is terrific, @agnivade!

> One additional thing that I have done from your PR is to update the upsert query as well. Please see https://github.com/mattermost/mattermost-server/commit/950bc1271d99a06823061d32a3b3e55cd0d92d92. Just ensuring that this was an oversight, and not something you intentionally avoided.

This goes back a ways, so I can't say with certainty what I intended then. But looking closely at the Upsert, it seems we also don't update the `PostId`. Since the `ChannelId` is effectively derived from the `PostId`, I'm wondering if these two fields were considered immutable. I'm not fundamentally opposed to allowing the sqlstore to mutate both of these values, but we should probably enforce the current assumptions at the application layer first to ensure no unexpected side effects. It might be simpler to keep the `ChannelId` immutable for the purposes of this PR.

**Comment by agnivade** [comment]
Results:

Bounded test for 10k users using the 12M data dump on a r6g.2xlarge graviton instance:

**Avg**
get_file_count-avg

**P99**
get_file_count-p99

Both average and P99 are showing improvements.

Query analysis:

```
agnivalt0db=> explain (analyze true, costs true, buffers true) select count(*) from fileinfo where fileinfo.deleteat=0 and fileinfo.postid in (select id from posts where channelid='nwn9435s6tgg7b8zfnb5ncut5w');
                                                                         QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=33999.39..33999.40 rows=1 width=8) (actual time=25.604..25.607 rows=1 loops=1)
   Buffers: shared hit=9298
   ->  Hash Join  (cost=1058.14..33999.39 rows=3 width=0) (actual time=7.211..25.546 rows=626 loops=1)
         Hash Cond: ((posts.id)::text = (fileinfo.postid)::text)
         Buffers: shared hit=9298
         ->  Bitmap Heap Scan on posts  (cost=339.67..33223.54 rows=9176 width=27) (actual time=4.124..18.619 rows=26275 loops=1)
               Recheck Cond: ((channelid)::text = 'nwn9435s6tgg7b8zfnb5ncut5w'::text)
               Heap Blocks: exact=8225
               Buffers: shared hit=8682
               ->  Bitmap Index Scan on idx_posts_channel_id_update_at  (cost=0.00..337.38 rows=9176 width=0) (actual time=3.150..3.150 rows=30831 loops=1)
                     Index Cond: ((channelid)::text = 'nwn9435s6tgg7b8zfnb5ncut5w'::text)
                     Buffers: shared hit=337
         ->  Hash  (cost=667.38..667.38 rows=4087 width=26) (actual time=2.558..2.559 rows=4159 loops=1)
               Buckets: 8192 (originally 4096)  Batches: 1 (originally 1)  Memory Usage: 304kB
               Buffers: shared hit=616
               ->  Seq Scan on fileinfo  (cost=0.00..667.38 rows=4087 width=26) (actual time=0.008..1.631 rows=4159 loops=1)
                     Filter: (deleteat = 0)
                     Rows Removed by Filter: 28
                     Buffers: shared hit=616
 Planning Time: 0.371 ms
 Execution Time: 25.934 ms
(21 rows)

agnivalt0db=>
agnivalt0db=> explain (analyze true, costs true, buffers true) select count(*) from fileinfo where fileinfo.deleteat=0 and fileinfo.channelid='nwn9435s6tgg7b8zfnb5ncut5w';
                                                  QUERY PLAN
--------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=679.21..679.22 rows=1 width=8) (actual time=2.211..2.211 rows=1 loops=1)
   Buffers: shared hit=616
   ->  Seq Scan on fileinfo  (cost=0.00..677.65 rows=623 width=0) (actual time=0.009..2.156 rows=626 loops=1)
         Filter: ((deleteat = 0) AND ((channelid)::text = 'nwn9435s6tgg7b8zfnb5ncut5w'::text))
         Rows Removed by Filter: 3570
         Buffers: shared hit=616
 Planning Time: 0.137 ms
 Execution Time: 2.233 ms
(8 rows)
```

This also shows that on paper, there is improvement also.

**Comment by agnivade** [comment]
@agarciamontoro , @lieut-data - PTAL again. All concerns are addressed.

**Review by lieut-data — APPROVED** [review]
:tada: -- just a few stylistic questions on the migrations, but nothing blocking.

File: db/migrations/mysql/000106_fileinfo_channelid.down.sql [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    'SELECT 1'
+));
+
+PREPARE removeIndexIfExists FROM @preparedStatement;
+EXECUTE removeIndexIfExists;
+DEALLOCATE PREPARE removeIndexIfExists;
+
+SET @preparedStatement = (SELECT IF(
+    EXISTS(
+        SELECT 1 FROM INFORMATION_SCHEMA.STATISTICS
```
[the remark this answers is not in the document]
Comment (agnivade) [reply db/migrations/mysql/000106_fileinfo_channelid.down.sql:18]: Not really, this was code that you wrote that I just lifted verbatim. :P Happy to make it consistent.

**Comment by agnivade** [comment]
Gentle ping @agarciamontoro

File: db/migrations/mysql/000106_fileinfo_channelid.down.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (agnivade) [reply db/migrations/mysql/000106_fileinfo_channelid.down.sql:18]: Went with `IF(EXISTS(` everywhere :)

File: db/migrations/mysql/000106_fileinfo_channelid.up.sql [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+SET @preparedStatement = (SELECT IF(
+    (
+        SELECT COUNT(*) FROM INFORMATION_SCHEMA.STATISTICS
+        WHERE table_name = 'FileInfo'
+        AND table_schema = DATABASE()
+        AND index_name = 'idx_fileinfo_channel_id_create_at'
+    ) > 0,
+    'SELECT 1',
+    'CREATE INDEX idx_fileinfo_channel_id_create_at ON FileInfo(ChannelId, CreateAt);'
+));
```
[the remark this answers is not in the document]
Comment (lieut-data) [reply db/migrations/mysql/000106_fileinfo_channelid.up.sql:30]: > Not that I am aware of. I didn't realize that prepared statements can have such a cognitive load. Do you think we should do something about it?

I'd love to see us verify that the database after down = the database before up, but it's also possible I'm just overthinking the potential failure points here. Another reason to prefer Postgres over MySQL :p

[the remark this answers is not in the document]
Comment (agnivade) [reply db/migrations/mysql/000106_fileinfo_channelid.up.sql:30]: Sometimes, we don't run destructive backward migrations. So it's not always a 100% match. :(

**Review by agarciamontoro — APPROVED** [review]
Flawless :ok_hand:

**Comment by agnivade** [comment]
Closing and re-opening again with monorepo.

**Pull request closed without merging**